### PR TITLE
proxy/ipvs: only get local addresses once per sync

### LINKED ISSUE #84906: performance enhance for kube-proxy/ipvs syncProxyRules
**What would you like to be added**:
performance enhance for kube-proxy/ipvs syncProxyRules

**Why is this needed**:
For kube-proxy/ipvs, syncProxyRules will take about 10s+ when the number of k8s Service with external ip > 1500+. It has those problem:
- The ipvs rules of endpoints change may delay 10s after a pod is terminated, within the period, all requests route to this endpoint will fail.
- The ipvs rules of Services change may delay 10s. It means that, a service may not be ready after created successfully until 10s later.

After some debug tests, I found that most of time spend on [utilproxy.IsLocalIP(externalIP)](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1103)
```
I1104 16:34:01.051111   14602 proxier.go:980] [debug] check externalIP open port: 4.5211ms
I1104 16:34:01.066563   14602 proxier.go:980] [debug] check externalIP open port: 6.349011ms
I1104 16:39:51.858951   11581 proxier.go:744] [debug] syncProxyRules took 12.045648906s
```

And fortunately, we could simply move the logic of geting net interface addrs from the for loop to enhance performance safely. After this change, the sync time could reduce from 12s to 730ms.
```
I1105 10:18:30.002215   31985 proxier.go:1002] [debug] check externalIP open port: 156ns
I1105 10:18:30.687643   31985 proxier.go:793] [debug] syncProxyRules, took 731.416349ms
```

I will push a PR soon for this.

### PR DESCRIPTION
**Pull request by njuicsgz**
**What type of PR is this?**
/kind bug
actually is enhancement

**What this PR does / why we need it**:

performance enhance for kube-proxy/ipvs syncProxyRules

**Which issue(s) this PR fixes**:
Fixes #84906

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by MrHohn** [comment]
Seems like there is merge conflict?
/ok-to-test
/cc @andrewsykim

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -899,6 +899,23 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
 // EntryInvalidErr indicates if an ipset entry is invalid or not
 const EntryInvalidErr = "error adding entry %s to ipset %s"
 
+func getLocalAddrs() (map[string]bool, error) {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: Make this a `map[string]struct{}` since we only check for existence

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -899,6 +899,23 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
...
+		return nil, err
+	}
+
+ 	localAddrs := map[string]bool{}
+	for i := range addrs {
+		intf, _, err := net.ParseCIDR(addrs[i].String())
+		if err != nil {
+			return nil, err
+		}
+		localAddrs[intf.String()] = true
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: Does this end up being the same equality check as `utilproxy.IsLocalIP`?

In `utilproxy.IsLocalIP` we have:
```
if net.ParseIP(ip).Equal(intf) {
    return true, nil
}
```

We're no longer doing the `ip.Equal` check and assuming the string values are the same.

**Review by andrewsykim — COMMENTED** [review]
> Seems like there is merge conflict? 

I don't think prow allows merge commits on PRs

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: I also see that `Equal` from `net.IP` considers an IPv4 address in IPv6 form equal to it's IPv4 form. 

```
// Equal reports whether ip and x are the same IP address.
// An IPv4 address and that same address in IPv6 form are
// considered to be equal.
func (ip IP) Equal(x IP) bool {
	if len(ip) == len(x) {
		return bytealg.Equal(ip, x)
	}
	if len(ip) == IPv4len && len(x) == IPv6len {
		return bytealg.Equal(x[0:12], v4InV6Prefix) && bytealg.Equal(ip, x[12:])
	}
	if len(ip) == IPv6len && len(x) == IPv4len {
		return bytealg.Equal(ip[0:12], v4InV6Prefix) && bytealg.Equal(ip[12:], x)
	}
	return false
}
```

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: git it

File: pkg/proxy/ipvs/proxier.go (hunk 2 above)
Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: yes, net.IP.Equal seems more safe for ipv6, I will rollback this compare logic.

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -899,6 +899,33 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
...
+		ip, _, err := net.ParseCIDR(addrs[i].String())
+		if err != nil {
+			return nil, err
+		}
+		localAddrs = append(localAddrs, ip)
+	}
+	return localAddrs, nil
+}
+
+func isLocalIP(ip string, addrs []net.IP) (bool) {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: This function seems more generic, wondering if this makes more sense:
```
func ipExists(ip net.IP, ips []net.IP) bool {
    ...
}
```

File: pkg/proxy/ipvs/proxier.go [hunk 4]
```diff
@@ -899,6 +899,33 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
...
+	obj := net.ParseIP(ip)
+	for i := range addrs {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: I think `for _, addr := range addrs` ends up reading better in the line below

File: pkg/proxy/ipvs/proxier.go [hunk 5]
```diff
@@ -899,6 +899,33 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
...
 const EntryInvalidErr = "error adding entry %s to ipset %s"
 
+func getLocalAddrs() ([]net.IP, error) {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+ 	var localAddrs []net.IP
+	for i := range addrs {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: `for _, addr := range addrs` would make the line below a bit easier to read

File: pkg/proxy/ipvs/proxier.go [hunk 6]
```diff
@@ -919,6 +946,12 @@ func (proxier *Proxier) syncProxyRules() {
 		klog.V(4).Infof("syncProxyRules took %v", time.Since(start))
 	}()
 
+	localAddrs, err := getLocalAddrs()
+	if err != nil {
+		klog.Errorf("Failed to sync proxy rules due to get local address error: %v", err)
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]:
```
klog.Errorf("Failed to get local addresses during proxy sync: %v", err)
```

**Review by andrewsykim — COMMENTED** [review]
Minor comments, PR looks good otherwise. Can we add more description to the commit message please? Maybe something like:
```
proxy/ipvs: only get local addresses once per sync
```

**Comment by andrewsykim** [comment]
/retitle proxy/ipvs: only get local addresses once per sync

**Comment by njuicsgz** [comment]
confused about the failure of 'pull-kubernetes-node-e2e-containerd', seems no effect to old logic

please /retest

File: pkg/proxy/ipvs/proxier.go (hunk 4 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: Don't think this was resolved yet?

**Review by andrewsykim — COMMENTED** [review]
/retest

Can you squash the commits and add a better description in the message please?

**Comment by njuicsgz** [comment]
> /retest
> 
> Can you squash the commits and add a better description in the message please?

done

File: pkg/proxy/ipvs/proxier.go [hunk 7]
```diff
@@ -919,6 +945,12 @@ func (proxier *Proxier) syncProxyRules() {
 		klog.V(4).Infof("syncProxyRules took %v", time.Since(start))
 	}()
 
+	localAddrs, err := getLocalAddrs()
+	if err != nil {
+		klog.Errorf("Failed to get local addresses during proxy sync: %v", err)
+		return
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: One change in behavior worth mentioning. Today, if we fail to get the local addresses, we just assume the external IP isn't part of local addresses. Now we log error and return early. Wondering if we need to preserve that behavior.

Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: Maybe don't return here, and on line 1135, do a check for `len(localAddrs)`. If `len(localAddrs) == 0`, assume externalIP is not local.

Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: thanks for catching this, we should continue syncing when get local adds failed for case that there is no external ip service.

File: pkg/proxy/ipvs/proxier.go [hunk 8]
```diff
@@ -899,6 +899,34 @@ func (proxier *Proxier) OnEndpointSlicesSynced() {
...
+
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return localAddrs, err
+	}
+
+	for _, addr := range addrs {
+		ip, _, err := net.ParseCIDR(addr.String())
+		if err != nil {
+			klog.Errorf("parse cidr failed when getting local addresses, err: %v", err)
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: This should still return error IMO

File: pkg/proxy/ipvs/proxier.go [hunk 9]
```diff
@@ -919,6 +947,11 @@ func (proxier *Proxier) syncProxyRules() {
 		klog.V(4).Infof("syncProxyRules took %v", time.Since(start))
 	}()
 
+	localAddrs, err := getLocalAddrs()
+	if err != nil {
+		klog.Errorf("Failed to get local addresses during proxy sync: %v, assuming external ip is not local", err)
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: We should actually omit the "assuming external IP is not local" part and log it later when we check external IPs.

File: pkg/proxy/ipvs/proxier.go [hunk 10]
```diff
@@ -1100,11 +1133,8 @@ func (proxier *Proxier) syncProxyRules() {
 
 		// Capture externalIPs.
 		for _, externalIP := range svcInfo.ExternalIPStrings() {
-			if local, err := utilproxy.IsLocalIP(externalIP); err != nil {
-				klog.Errorf("can't determine if IP is local, assuming not: %v", err)
-				// We do not start listening on SCTP ports, according to our agreement in the
-				// SCTP support KEP
-			} else if local && (svcInfo.Protocol() != v1.ProtocolSCTP) {
+			if ipExists(net.ParseIP(externalIP), localAddrs) && (svcInfo.Protocol() != v1.ProtocolSCTP) {
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]:
```
if len(localAddrs) ==.0 {
      klog.Errorf("couldn't find any local IPs, assuming %s is not local", externalIP)
} else if (svcInfo.Protocol() != v1.ProtocolSCTP) && ipExists(net.ParseIP(externalIP), localAddrs) {
    ...
}
```

File: pkg/proxy/ipvs/proxier.go (hunk 8 above)
Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: If there is 100 adds, 1 of them parsed error, we could still get 99 correctly. If return error directly, we may get empty local addrs, so any external ip is assuming not local ip. How do you think about this?

File: pkg/proxy/ipvs/proxier.go (hunk 9 above)
Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: It depends on above comments. If we permit returning partly correct addrs, it is hard to distinguish if addrs is correct or partly correct.

File: pkg/proxy/ipvs/proxier.go (hunk 10 above)
Comment (njuicsgz) [reply pkg/proxy/ipvs/proxier.go]: it make sense, I will fix it later

File: pkg/proxy/ipvs/proxier.go (hunk 8 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: I think we can consider not returning err here but should be in another PR.

**Comment by andrewsykim** [comment]
/approve
/lgtm

@thockin @robscott we should consider a similar optimization for the iptables proxy

**Comment by andrewsykim** [comment]
Thank you @njuicsgz!

**Comment by njuicsgz** [comment]
> /approve
> /lgtm
> 
> @thockin @robscott we should consider a similar optimization for the iptables proxy, more details in #84906

https://github.com/kubernetes/kubernetes/pull/85433 try to enhance similar performance issue in userspace mode, which seems LGTM. Could you have a look?

**Pull request merged**